metrics-operator/controllers/analysis/controller.go

```diff
@@ -0,0 +1,167 @@
+//+kubebuilder:rbac:groups=metrics.keptn.sh,resources=analysisvaluetemplates,verbs=get;list;watch;
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+// For more details, check Reconcile and its AnalysisResult here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.13.0/pkg/reconcile
+func (a *AnalysisReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	a.Log.Info("Reconciling Analysis")
+	analysis := &metricsapi.Analysis{}
+
+	//retrieve analysis
+	if err := a.Client.Get(ctx, req.NamespacedName, analysis); err != nil {
+		if errors.IsNotFound(err) {
+			// taking down all associated K8s resources is handled by K8s
+			a.Log.Info("Analysis resource not found. Ignoring since object must be deleted")
+			return ctrl.Result{}, nil
+		}
+		a.Log.Error(err, "Failed to get the Analysis")
+		return ctrl.Result{}, err
+	}
+
+	//find AnalysisDefinition to have the collection of Objectives
+	analysisDef := &metricsapi.AnalysisDefinition{}
+	if analysis.Spec.AnalysisDefinition.Namespace == "" {
+		analysis.Spec.AnalysisDefinition.Namespace = a.Namespace
+	}
+	err := a.Client.Get(ctx,
+		types.NamespacedName{
+			Name:      analysis.Spec.AnalysisDefinition.Name,
+			Namespace: analysis.Spec.AnalysisDefinition.Namespace},
+		analysisDef,
+	)
+
+	if err != nil {
+		if errors.IsNotFound(err) {
+			a.Log.Info(err.Error() + ", ignoring error since object must be deleted")
+			return ctrl.Result{Requeue: true, RequeueAfter: 10 * time.Second}, nil
+		}
+		a.Log.Error(err, "Failed to retrieve the AnalysisDefinition")
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, nil
+	}
+
+	var done map[string]metricsapi.ProviderResult
+	todo := analysisDef.Spec.Objectives
+	if analysis.Status.StoredValues != nil {
+		todo, done = extractMissingObjectives(analysisDef.Spec.Objectives, analysis.Status.StoredValues)
+		if len(todo) == 0 {
+			return ctrl.Result{}, nil
+		}
+	}
+
+	//create multiple workers handling the Objectives
+	childCtx, wp := a.NewWorkersPoolFactory(ctx, analysis, todo, a.MaxWorkers, a.Client, a.Log, a.Namespace)
+
+	res, err := wp.DispatchAndCollect(childCtx)
+	if err != nil {
+		a.Log.Error(err, "Failed to collect all values required for the Analysis, caching collected values")
+		analysis.Status.StoredValues = res
+		err = a.updateStatus(ctx, analysis)
+		return ctrl.Result{RequeueAfter: 10 * time.Second}, err
+	}
+
+	maps.Copy(res, done)
+
+	err = a.evaluateObjectives(ctx, res, analysisDef, analysis)
+
+	return ctrl.Result{}, err
+}
+
+func (a *AnalysisReconciler) evaluateObjectives(ctx context.Context, res map[string]metricsapi.ProviderResult, analysisDef *metricsapi.AnalysisDefinition, analysis *metricsapi.Analysis) error {
+	eval := a.Evaluate(res, analysisDef)
+	analysisResultJSON, err := json.Marshal(eval)
+	if err != nil {
+		a.Log.Error(err, "Could not marshal status")
```

Comment:
We should return here, if we cannot unmarshall the result, the re is no need to store it

Reply:
adapted, but changed to just not set the raw status in case of an error here, but proceed with setting pass/warning properties. In reality, we will never get an error from the marshal method anyway

Reply:
sure, thanks :)